File: day23/ProductExceptSelf.java
import java.util.*;

public class ProductExceptSelf {

    public static int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] output = new int[n];

        // Prefix pass
        int prefix = 1;
        for (int i = 0; i < n; i++) {
            output[i] = prefix;
            prefix *= nums[i];
        }

        // Suffix pass
        int suffix = 1;
        for (int i = n - 1; i >= 0; i--) {
            output[i] *= suffix;
            suffix *= nums[i];
        }

        return output;
    }

    public static void main(String[] args) {
        Random rand = new Random();
        int length = rand.nextInt(6) + 5; // length 5 to 10
        int[] nums = new int[length];
        for (int i = 0; i < length; i++) {
            nums[i] = rand.nextInt(10) + 1; // values 1 to 10
        }

        System.out.println("Input: " + Arrays.toString(nums));
        int[] result = productExceptSelf(nums);
        System.out.println("Output: " + Arrays.toString(result));
    }
}

File: day23/product_except_self.rs
use rand::random_range;

fn product_except_self(nums: &[i32]) -> Vec<i32> {
    let n = nums.len();
    let mut output = vec![1; n];

    // Prefix pass
    let mut prefix = 1;
    for i in 0..n {
        output[i] = prefix;
        prefix *= nums[i];
    }

    // Suffix pass
    let mut suffix = 1;
    for i in (0..n).rev() {
        output[i] *= suffix;
        suffix *= nums[i];
    }

    output
}

fn main() {
    let len = random_range(5..=10);
    let nums: Vec<i32> = (0..len).map(|_| random_range(1..=10)).collect();

    println!("Input: {:?}", nums);
    let result = product_except_self(&nums);
    println!("Output: {:?}", result);
}
